python/lib/core/dmod/core/common/tasks.py

```diff
@@ -0,0 +1,180 @@
+"""
+The maximum amount of attempts to wait on a task to finish.
+
+A task may not finish within the given time, so this lets the function extend its waiting period while making sure 
+that it completes without taking too much time. Increase to poll more times (useful when you poll frequently on 
+possibly long running tasks) and decrease to exit without completion sooner
+"""
+
+TASK_MESSAGE = typing.Union[str, typing.Callable[[asyncio.Task], typing.Union[typing.Coroutine, str]]]
+"""
+Either a string or a function that will generate a string that will be logged in relation to an asynchronous task
+"""
+
+TASK_EXCEPTION_MESSAGE = typing.Union[
+    str,
+    typing.Callable[
+        [asyncio.Task, typing.Optional[BaseException]],
+        typing.Union[
+            typing.Coroutine,
+            str
+        ]
+    ]
+]
+"""
+Either a string or a function that will generate a string that will be logged as an exception message in relation 
+to an asynchronous task
+"""
+
+
+async def wait_on_task(task: asyncio.Task, wait_seconds: int = None, maximum_times_to_poll: int = None) -> bool:
+    """
+    Wait for a task to finish
+
+    Will only wait a limited number of times before moving on
+
+    Note: Setting maximum_times_to_poll to a number less than 1 will poll until the task has been marked as completed
+
+    Args:
+        task: The task to poll
+        wait_seconds: The number of seconds to wait between polling
+        maximum_times_to_poll: The maximum number of times to poll the running task
+
+    Returns:
+        Whether the task completed during or prior to the polling period
+    """
+    if not isinstance(wait_seconds, typing.SupportsInt):
+        logging.warning(
+            f"'{get_current_function_name(parent_name=True)}' did not supply a valid amount of seconds to "
+            f"wait for polling. Defaulting to {DEFAULT_TASK_WAIT_SECONDS} seconds"
+        )
+        wait_seconds = DEFAULT_TASK_WAIT_SECONDS
+    elif wait_seconds is None:
+        wait_seconds = DEFAULT_TASK_WAIT_SECONDS
+
+    if not isinstance(maximum_times_to_poll, typing.SupportsInt):
+        logging.warning(
+            f"'{get_current_function_name(parent_name=True)}' did not supply a valid number of times to poll. "
+            f"Defaulting to {DEFAULT_MAXIMUM_POLLS}"
+        )
+        maximum_times_to_poll = DEFAULT_MAXIMUM_POLLS
+    elif maximum_times_to_poll is None:
+        maximum_times_to_poll = DEFAULT_MAXIMUM_POLLS
+
+    # Poll the task until it is deemed completed
+    # Just move on if it takes too long
+    wait_count = 0
+    while not task.done() and (wait_count < maximum_times_to_poll or maximum_times_to_poll < 0):
+        # Wait for the task to finish just in case it has its own asyncio sleep within (possible for producers)
+        await asyncio.sleep(wait_seconds)
+
+        if maximum_times_to_poll > 0:
+            wait_count += 1
+
+    return task.done() or task.cancelled()
+
+
+async def cancel_task(
```

Comment:
What are your thoughts @christophertubbs on returning a boolean that indicates success from this coroutine?

Reply:
I'm having it return a new `CancelResults` type. That way it'll give that boolean _and_ say what failed to cancel. That function is used by `cancel_tasks`, so I imagine that should return something as well, but a tuple of booleans isn't particularly enlightening.

Reply:
Hmm, im not seeing where `CancelResults` is being returned? Did that get committed?